### ISSUE TITLE
TensorRT deployment

### ISSUE THREAD

**Issue by davidkflau** [MIC-DKFZ/nnUNet#2706]
Hi,

I'm trying to optimize the inference speed by converting the model to tensorRT using pytorch-tensorrt. I added below code to the `predict_sliding_window_return_logits` and saved the tensorRT model, but the inference speed of the tensorRT model is 10x slower than the pytorch model. Does anyone know how to convert the model to tensorRT properly?

 input = torch.randn((1, 5, 20, 320, 256)).cuda()
trt_gm = torch_tensorrt.compile(self.network, ir="dynamo", inputs=[input], backend="torch_tensorrt", dynamic=False,
                        options={"truncate_long_and_double": True,
                                "enabled_precisions": {torch.float,torch.half},
                                "debug": False,
                                "optimization_level": 5})

torch_tensorrt.save(trt_gm, "model.ep", inputs=[input])

**Comment by FabianIsensee** [comment]
Hey unfortunately we have not yet invested effort into this but I have it on my todo list. Have you been able to make progress in the meantime?

**Comment by MaximoRdz** [comment]
Hello, it is difficult to reproduce the exact issue as I don't know which model configuration, GPU model, and library versions you're using.

Nonetheless, as pointed out in the NVIDIA TensorRT documentation, precision is one of the main factors affecting inference performance ([docs](https://docs.pytorch.org/TensorRT/user_guide/performance_tuning.html#using-the-right-precision)). From what I can see, you used `input = torch.randn((1, 5, 20, 320, 256)).cuda()`; of course, this is just the dummy input, but it may reflect the same issue in your source code. By default, `torch.randn` creates a `torch.float32` tensor, and the nnUNet model is also in `torch.float32`. Because of this, simply setting `enabled_precisions` to allow both `half` and `float` does not necessarily reproduce the mixed-precision behavior used by nnUNet. With recent versions of Torch-TensorRT/TensorRT, lower-precision execution needs to be explicitly enabled, for example through autocast. To check how TensorRT is partitioning your model, you can add `dryrun=True` to `torch_tensorrt.compile()` and inspect which operations are being compiled by TensorRT and to which precision.

And, answering the original question: in such a case, the generated engine is slower because during inference `nnUNet` uses autocast ([here](https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunetv2/inference/predict_from_raw_data.py#L681)) whenever possible, effectively working in `torch.half` most of the time, which will inevitably outperform a TensorRT fp32 engine.

```python
    @torch.inference_mode()
    def predict_sliding_window_return_logits(self, input_image: torch.Tensor) \
            -> Union[np.ndarray, torch.Tensor]:
        [. . .]
        with torch.autocast(self.device.type, enabled=True) if self.device.type == 'cuda' else dummy_context():
            [. . .]
```

The plot below shows exactly this: `github-issue` (the reported setup) and `trt-solution-fp32` (forcing compilation to work only with precision fp32) sit roughly an order of magnitude above every other configuration, because both are effectively running fp32 under the hood.

<img width="1650" height="900" alt="Image" src="https://github.com/user-attachments/assets/f6494a8d-89b2-4620-b3a8-d69537232755" />

## Proposed Solution

This is the easiest solution I could find, and it best aligns with nnUNet's "avoid complexity" philosophy: enable autocast during compilation to mimic `nnUNetPredictor`'s behavior and get the same precision. (It would be preferable to control the precision exactly by setting it to fp16, but we can keep it general to all GPUs in case the compiler detects during autocasting that fp16 is actually a bad option.) The `require_full_compilation` flag is useful to avoid silent graph breaks; for now, all nnUNet configurations are fully covered by TensorRT anyways.

- **compile and export**
```python
trt_gm = torch_tensorrt.compile(
    network,
    ir="dynamo",
    inputs=[input_data],
    backend="torch_tensorrt",
    dynamic=False,
    enable_autocast=True,
    autocast_low_precision_type=torch.float16,
    require_full_compilation=True,
)

torch_tensorrt.save(trt_gm, output_filepath, inputs=[input_data])
```

- **loading**
```python
network = torch.export.load(output_filepath).module()
```

### Results summary

Sadly, autocast TensorRT is not always the fastest option, in several configurations it's actually beaten by plain `torch.compile` (see below):

<img width="2520" height="900" alt="Image" src="https://github.com/user-attachments/assets/ab591336-d20d-4c61-9624-8cee92821e49" />

## Caveats

TensorRT is an actively developed project, and some operations aren't yet perfectly optimized by the compiler. I'm talking specifically about skip connections. At least from what I've been able to find, TensorRT doesn't have a specific way of optimizing skip connections as a whole. Meaning: in a UNet model, TensorRT optimizes `encoder -> bottleneck -> decoder` in order and independently, so the optimal policy for the encoder might not align with the decoder's (and in my experiments, it did not). For increasingly larger inputs, performance ends up heavily dominated by memory bandwidth during the skip connections' `torch.cat` operations, plus unnecessary reformatting and moves of the skipped features to match the decoder's upcoming features (NHWC-type format, etc.). These operations clearly don't scale well, and cause models with large input sizes compiled with TensorRT to perform worse than plain PyTorch (unless the GPU architecture bandwidth is good enough to tolerate this as in the case of DGX GB100, it would be interesting to test in A100, etc).

One workaround, until the TensorRT compiler gets smarter about this, would be to replace the costly concatenation of the two tensors followed by a convolution, with the sum of two properly configured partial convolutions (mathematically equivalent, no retraining needed):

$$conv(cat(features, skips)) = conv1(features) + conv2(skips)$$

But then again, this introduces the overhead of more kernel launches, and while it performs better than plain eager PyTorch, this custom TensorRT solution is often still slower than the simpler `torch.compile(network)` for these large input tensors.

## Experiments

As this topic aligns with my thesis, I performed some experiments to check. In particular, I benchmarked a single forward pass of the model (eager mode, `torch.compile`, TensorRT, etc.) on 3 available GPUs and 2 datasets: one public, and our private one.
### Setup

| Specification  | DGX Spark   | A40        | L40S        |
| -------------- | ----------- | ---------- | ----------- |
| GPU            | NVIDIA GB10 | NVIDIA A40 | NVIDIA L40S |
| Driver         | 580.173.02  | 595.71.05  | 575.57.08   |
| CUDA           | 13.0        | 13.2       | 12.9        |
| Python         | 3.12.3      | 3.12.12    | 3.12        |
| PyTorch        | 2.13.0      | 2.13.0     | 2.8.0       |
| TensorRT       | 11.0.0.114  | 11.0.0.114 | —           |
| Torch-TensorRT | 2.13        | 2.13.0     | 2.8.0+cu129 |
| nnUNet         | 0e49508     | 0e49508    | 0e49508     |

### Datasets
- Automated Cardiac Diagnosis Challenge (ACDC Dataset027) — [kaggle](https://www.kaggle.com/datasets/anhoangvo/acdc-dataset/data)
        - Tested with configurations 2d and 3d_fullres, planner: `nnUNetPlannerResEncL`
    - 2D configuration patch size: `(256, 10)`
    - 3D_fullres configuration patch size: `(256, 256, 10)`
    
- Bone tumor segmentation (private dataset Dataset306), labels: 0 = background, 1 = tumor, CT scans of `(N, 512, 512)`. Used an old model trained on `nnUNetPlans`, no residual encoder.
    - 2D configuration patch size: `(512, 512)`
    - 3D_fullres configuration patch size: `(48, 192, 192)`
## Results
Several configuratons tested:
* `github-issue`: Implementation/configuration from the GitHub issue.
* `pytorch`: Standard PyTorch nnUNet inference implementation (baseline, normalized to 1.00×).
* `pytorch-compile`: PyTorch model compiled with torch.compile (no NHWC or split-conv modifications).
* `pytorch-compile-nhwc-split-conv`: torch.compile using NHWC layout and split-convolution transformation on skip connections.
* `pytorch-nhwc`: Standard PyTorch inference using NHWC memory layout (no torch.compile or split-conv).
* `pytorch-nhwc-split-conv`: PyTorch with NHWC layout and split-convolution, without torch.compile.
* `trt-solution-fp32`: Proposed TensorRT solution running in full FP32 precision.
* `trt-solution-autocast`: TensorRT solution using autocast for automatic lower-precision selection.
* `trt-solution-fp16`: TensorRT solution explicitly configured for FP16 precision.
* `trt-solution-fp16-nhwc`: TensorRT FP16 configuration combined with NHWC memory layout.
* `trt-solution-fp16-nhwc-split-conv`: TensorRT FP16 + NHWC layout with split-convolution to optimize skip-connection concats.

### 1. 2D configuration

**ACDC (Dataset027)**

TensorRT compiled engine performs really well for 2d as the task is compute-bound and not really bandwidth bound the fused kernels and other optimizations dominate over `torch.compile` (patch size of ~ 5.1 KB for fp16)

<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/b37025ce-a036-4c15-841c-728176cb977c" />
<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/01859998-dd67-427b-8cf0-82bbd38a3b76" />
<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/72e585b4-ad86-41fd-a751-7f060650dbf0" />

**BONES (Dataset306)**

Again we see some improvements by using tensorRT but the larger input size already appears (patch size ~ 0.5 MB) particularly bad for L40S

<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/1fbbd762-a09a-4147-b518-6367b8f98016" />
<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/5e7a9118-c4f9-4948-853d-8d68fcd6e9be" />
<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/0d5f379a-a950-4d80-a777-c271879bc4b6" />


### 2. 3D_fullres configuration

**ACDC (Dataset027)**

`torch.compile` is better for GPUs with worst bandwidth where the bad compilation of skip connections effect dominates. 

<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/ffb1556b-65f7-422b-b5b7-00b3cd506f40" />
<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/04203fac-eff7-4594-a8bf-daff33dc708a" />
<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/fcbdfbe1-3e71-4040-9e79-679ac4660c61" />


**BONES (Dataset306)**

<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/b6e50fc9-b607-4122-bc69-1167ccdc9fb3" />
<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/1e7ec4dd-daf4-4d0b-bf74-1c8edb724435" />
<img width="675" height="450" alt="Image" src="https://github.com/user-attachments/assets/0767397c-7dfa-42d5-8ad4-39cf5e215265" />

### 3. TensorRT explosion on 3D_fullres skip connections

Skip-connection concatenation becomes a bottleneck on larger inputs. Below is the profiling evidence on the A40.

**Dataset027 dataset, 3D fullres**

<img width="9292" height="1440" alt="Image" src="https://github.com/user-attachments/assets/332cc9f5-5850-4c91-bcfa-1f04f0feb580" />

**Dataset306 Bone tumor dataset, 3D fullres**

<img width="2841" height="1440" alt="Image" src="https://github.com/user-attachments/assets/9265a8b0-dd8c-4ac3-8018-9d6039ce3f7e" />

**Naive solution** Bone tumor dataset, 3D fullres, replacing concatenation with two convolutions:

<img width="3302" height="1440" alt="Image" src="https://github.com/user-attachments/assets/4d9cb19a-ef11-4a3b-81c7-67074b1c0433" />

## Misc Comments

TensorRT offers some parameters to tune the final compiled engine:
- tiling: recommended for convolution-intensive networks
- workspace
- aux streams
- ...

but in order to keep nnUNet's simplicity philosophy, I believe enabling autocast as the only setting clearly provides all the benefit without introducing the hassle of adapting the compilation to every possible GPU model a user might have.
